added reporting hook

Galileo1 opened the pull request:
Added hook.js for cucumber json&html reporting. Also created a report folder when tractor is initialized.  And hook.js file will be copied  in support folder.

Commit:
adding reporting hook

added hook.js for cucumber json&html reporting

phenomnomnominal commented on package.json:
```diff
@@ -21,14 +21,15 @@
     "browserify-shim": "3.8.10",
     "chai": "2.3.0",
     "chai-as-promised": "5.1.0",
+    "cucumber-html-report": "^0.2.5",
```
Change this to the specific version (kill the ^)

phenomnomnominal commented on server/cli/init/base-file-sources/hooks.js:
```diff
@@ -0,0 +1,41 @@
+'use strict';
```
A few things here..
- Order the imports alphabetically, easier to read.
- Import bluebird here and promisify `fs`

phenomnomnominal commented on server/cli/init/base-file-sources/hooks.js:
```diff
@@ -0,0 +1,41 @@
+var jsonFormatter = cucumber.Listener.JsonFormatter();
+
+
+var tractorReportHook = function () {
```
No need for the `var`, just export the function directly.

I think the exported function should only do two things, create the reporter, and register the listener:

``` javascript
module.exports = function () {
    var reporter = createReporter();
    this.registerListener(reporter);
};
```

Galileo1 replied:
I believe exporting the json formatter listener(cucumber.Listener.JsonFormatter) will do the job.

phenomnomnominal commented on server/cli/init/base-file-sources/hooks.js:
```diff
@@ -0,0 +1,41 @@
+                createHtmlReport(cucumberReport);
+            }
+        });
+    };
```
This would all be moved out to the `createReporter` function. 
- the require should be moved to the top of the file.
- the comments are unnecessary
- fs.writeFile should be fs.writeFileAsync (look at server/cli/init/create-base-test-files.js for an example)

I'm thinking that the reports should be timestamped somehow? It'll mean they might have to be cleared out semi-regularly, but it'd be nice to be able to identify unique test runs

Galileo1 replied:
I have timestamped the reports (using moment.js).

phenomnomnominal commented on server/cli/init/create-base-test-files.js:
```diff
@@ -20,6 +20,8 @@ function createBaseTestFiles (testDirectoryPath) {
     .catch(TractorError, error => logNotCopying(error))
     .then(() => createProtractorConf(testDirectoryPath))
+    .catch(TractorError, error => logNotCopying(error))
+    .then(() => createHooksFile(join(testDirectoryPath, constants.SUPPORT_DIR)))
```
- pull `join(testDirectoryPath, constants.SUPPORT_DIR)` out to a variable

Commit:
incorporating review comments

Commit:
some more stuff

phenomnomnominal commented on server/cli/init/base-file-sources/hooks.js:
```diff
@@ -1,41 +1,44 @@
+var Promise = require('bluebird');
+
+var fs = Promise.promisifyAll(require('fs'));
  /* eslint-disable new-cap */
```
This lint override shouldn't be necessary cause `new` isn't being used

Galileo1 replied:
I believe the lint override is needed as "By default both newIsCap and capIsNew options are set to true."

phenomnomnominal commented on server/cli/init/base-file-sources/hooks.js:
```diff
@@ -1,41 +1,44 @@
+
+var fs = Promise.promisifyAll(require('fs'));
  /* eslint-disable new-cap */
 var jsonFormatter = cucumber.Listener.JsonFormatter();
```
I still don't like that this is just global to the file, it'd be cleaner to have a function that makes this, adds the log function, and encapsulates the functions that save the reports

phenomnomnominal commented on server/cli/init/base-file-sources/hooks.js:
```diff
@@ -1,41 +1,44 @@
-    };
+function createFileName (file, extension) {
+    /* eslint-disable prefer-template */
+    return file + moment().format('YYYY-MM-DD_HH-mm-ss') + '.' + extension;
```
Move the date format out to a variable `REPORT_DATE_FORMAT` or something.

Commit:
added the logger and encapsulate all () => ()

phenomnomnominal commented on server/cli/init/base-file-sources/hooks.js:
```diff
@@ -1,44 +1,50 @@
 'use strict';
 
-/* eslint-disable no-var, prefer-arrow-callback */
+/* eslint-disable no-var*/
```
Add this back and change back from arrows to function expressions, we can't guarantee this will be run in a version of Node that supports them. We could transpile this, but that's probably a whole different thing. otherwise it looks all good

phenomnomnominal commented on server/cli/init/base-file-sources/hooks.js:
```diff
@@ -1,44 +1,50 @@
+    /* eslint-disable new-cap */
+    var jsonFormatter = cucumber.Listener.JsonFormatter();
+    var outputDir = path.join(__dirname, '../', constants.REPORT_DIR)
+    jsonFormatter.log = content => jsonReportWriter(content);
```
This can just be `jsonFormatter.log = jsonReportWriter;` shouldn't need to worry about binding or anything,

Galileo1 replied:
hey , If I don't bind it , log function will write the content to the console.However to write it into a file , I need to bind it into a function. In our case jsonFormatter.log = function(content) {  and write content to the file}.

phenomnomnominal replied:
Nah sorry, I meant as in Function.prototype.bind. 

`jsonFormatter.log = content => jsonReportWriter(content);` is equivalent to `jsonFormatter.log = jsonReportWriter;` here because we don't use `this`.

Commit:
removing the lexical this binding..

make sense -> functions = objects = first class citizens in js and can
be passed around as if we are passing value.

Commit:
added back disabled arrow callback..
